### PR TITLE
Add BitsWriter write callback

### PR DESCRIPTION
**Pull request by barbashov**
Adding callback to BitsWriter which is called after every write.

I need this to calculate CRC32 while muxing PSI packets.

### CODE REVIEW

File: binary.go [hunk 1]
```diff
@@ -16,19 +16,26 @@ type BitsWriter struct {
...
 
+type BitsWriterWriteCallback func([]byte)
+
 // BitsWriterOptions represents BitsWriter options
 type BitsWriterOptions struct {
 	ByteOrder binary.ByteOrder
 	Writer    io.Writer
+
+	// WriteCallback is called every time when
+	WriteCallback BitsWriterWriteCallback
```
Comment (asticode) [inline binary.go:28]: Could you move this attribute above `Writer` and make sure the comment is complete?

File: binary.go [hunk 2]
```diff
@@ -16,19 +16,26 @@ type BitsWriter struct {
...
 }
 
 // NewBitsWriter creates a new BitsWriter
 func NewBitsWriter(o BitsWriterOptions) (w *BitsWriter) {
 	w = &BitsWriter{
 		bo:      o.ByteOrder,
 		w:       o.Writer,
+		writeCb: o.WriteCallback,
 		bsCache: make([]byte, 1),
```
Comment (asticode) [inline binary.go]: Could you move `bsCache` below `bo`?

File: binary.go [hunk 3]
```diff
@@ -111,16 +122,21 @@ func (w *BitsWriter) writeFullInt(in uint64, len int) error {
 func (w *BitsWriter) writeFullByte(b byte) error {
 	if w.cacheLen == 0 {
 		w.bsCache[0] = b
-		_, err := w.w.Write(w.bsCache)
-		return err
-	}
+		if _, err := w.w.Write(w.bsCache); err != nil {
```
Comment (asticode) [inline binary.go]: I'd rather create

```go
func (w *BitsWriter) write() error {
  if _, err := w.w.Write(w.bsCache); err != nil {
    return err
  }
  if w.writeCb != nil {
    w.writeCb(w.bsCache)
  }  
}
```

that gets called anytime a `write` is needed.

What do you think?

**Review by asticode — CHANGES_REQUESTED** [review]

File: binary.go (hunk 1 above)
Comment (barbashov) [reply binary.go:28]: whoa how come i haven't completed that comment :) done

File: binary.go (hunk 2 above)
Comment (barbashov) [reply binary.go]: sure, done

File: binary.go (hunk 3 above)
Comment (barbashov) [reply binary.go]: good call! that simplified the code

**Comment by asticode** [comment]
FYI I've created a `v0.18.0` tag.

Cheers

**Comment by barbashov** [comment]
Thanks!

**Pull request merged**